aws/resource_aws_storagegateway_fsx_associate_file_system.go

```diff
@@ -0,0 +1,304 @@
+				Required:  true,
+				Sensitive: true,
+				ValidateFunc: validation.All(
+					validation.StringMatch(regexp.MustCompile(`^[ -~]+$`), ""),
+					validation.StringLenBetween(1, 1024),
+				),
+			},
+			"tags":     tagsSchema(),
+			"tags_all": tagsSchemaComputed(),
+			"username": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.All(
+					validation.StringMatch(regexp.MustCompile(`^\w[\w\.\- ]*$`), ""),
+					validation.StringLenBetween(1, 1024),
+				),
+			},
+		},
+	}
+}
+
+func resourceAwsStorageGatewayFsxAssociateFileSystemCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).storagegatewayconn
+	defaultTagsConfig := meta.(*AWSClient).DefaultTagsConfig
+	tags := defaultTagsConfig.MergeTags(keyvaluetags.New(d.Get("tags").(map[string]interface{})))
+
+	input := &storagegateway.AssociateFileSystemInput{
+		ClientToken: aws.String(resource.UniqueId()),
+		GatewayARN:  aws.String(d.Get("gateway_arn").(string)),
+		LocationARN: aws.String(d.Get("location_arn").(string)),
+		Password:    aws.String(d.Get("password").(string)),
+		Tags:        tags.IgnoreAws().StoragegatewayTags(),
+		UserName:    aws.String(d.Get("username").(string)),
+	}
+
+	if v, ok := d.GetOk("audit_destination_arn"); ok {
+		input.AuditDestinationARN = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("cache_attributes"); ok {
+		input.CacheAttributes = expandStorageGatewayFsxAssociateFileSystemCacheAttributes(v.([]interface{}))
+	}
+
+	log.Printf("[DEBUG] Associating File System to Storage Gateway: %s", input)
+	output, err := conn.AssociateFileSystem(input)
+	if err != nil {
+		if tfawserr.ErrCodeEquals(err, storagegateway.ErrCodeInvalidGatewayRequestException) {
+			var igrex *storagegateway.InvalidGatewayRequestException
+			if ok := errors.As(err, &igrex); ok {
+				if err := igrex.Error_; err != nil {
+					if aws.StringValue(err.ErrorCode) == "FileSystemAssociationNotFound" {
+						log.Printf("[WARN] FSX File System %q not found, removing from state", d.Id())
+						d.SetId("")
+						return nil
+					}
+				}
+			}
+		}
+
+		return fmt.Errorf("Error associating file system to storage gateway: %w", err)
+	}
+
+	d.SetId(aws.StringValue(output.FileSystemAssociationARN))
+	log.Printf("[INFO] Storage Gateway FSx File System Association ID: %s", d.Id())
+
+	if _, err = waiter.FsxFileSystemAvailable(conn, d.Id(), d.Timeout(schema.TimeoutCreate)); err != nil {
+		return fmt.Errorf("error waiting for Storage Gateway FSx File System Association (%q) to be Available: %w", d.Id(), err)
+	}
+
+	return resourceAwsStorageGatewayFsxAssociateFileSystemRead(d, meta)
+}
+
+func resourceAwsStorageGatewayFsxAssociateFileSystemRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).storagegatewayconn
+	defaultTagsConfig := meta.(*AWSClient).DefaultTagsConfig
+	ignoreTagsConfig := meta.(*AWSClient).IgnoreTagsConfig
+
+	input := &storagegateway.DescribeFileSystemAssociationsInput{
```

Comment:
let extract the lookup logic to the relevant `finder` package an reuse this in acctests Exists and Destory functions

Reply:
Great idea. Thanks for that pointer! Updated.